### ISSUE TITLE
Aimless sandworm ignores gravity

### ISSUE THREAD

**Issue by Quuxplusone** [zenorogue/hyperrogue#97]
By an "aimless" sandworm I mean one which can't reach the player and therefore heads off to infinity in whatever direction it's currently facing.

#94 was about aimless sandworms pathing into not-yet-generated territory.

The present issue is about aimless sandworms heading straight up the Ivory Tower like a rocket ship.

![](https://i.imgur.com/HaZsuKw.png)

Here I jumped up onto a ladder to make the worm think it couldn't reach me; and then waited until the worm's head was out of sight; and then was delighted to find that the worm's body supported my weight and I was able to climb it to infinity. (Eventually I got into view of the head, at which point the worm turned around and started heading toward me again.)

----

Not really related, but: I was surprised to find that I can stand atop a worm tail in the Ivory Tower, but apparently I cannot stand atop a Metal Beast in its "waiting" state.

**Comment by ammongit** [comment]
I believe this is because worms, as multi-tile creatures, are supported by earlier tiles and can thus climb upwards forever, even if its tail is in the air.

I feel like this behavior should be changed. Instead, the body segments shouldn't count as platforms for the sandworm head, so it needs a real platform or falls. Or perhaps just causing them to avoid entering gravity lands?

**Comment by zenorogue** [comment]
Indeed, worms count as platforms, and they can in particular use their own body to support themselves.

The idea about climbing even if the tail is in the air looks strange, but I think this is fine too -- an alternate solution would be to let them fall if no body part is supported (I believe the game Snakebird works like this), but this introduces a funny result due to the hyperbolic geometry -- if you have, say, 10 worm tiles on the same level, the level below will be shorter, so these 10 worm tiles could not actually fall! Physically, this means that the gravity system used makes "wide" objects partially support themselves. Presumably the worms can use such non-Euclidean quirks to climb forever.

Regarding standing atop a Metal Beast -- this situation comes up frequently in the Dungeon, and I have decided that you can only use multi-tile creatures (worms and ivies) for climbing -- IIRC the reason was that it seems better to have clear loss condition (you know you have lost in the Dungeon if there is no way up, stacking skeletons weirdly will not help you).

**Comment by Quuxplusone** [comment]
Those are all good points. Still... is it possible to prevent sandworms from entering gravity lands? or is that a fool's errand? (I.e., emmiegit's second idea.)

Alternatively, couldn't you constrain the worm's head to always respect gravity? I mean that when the worm is choosing which way to go, it should be limited to the same directions that the Rogue would be in that situation. IIUC, the problem is that the worm head thinks it can "stand on top of itself" to climb upward, whereas the Rogue, being single-tile, cannot "stand on top of itself" and therefore naturally respects gravity. I wonder if this is just a matter of saying that worm _tails_ provide support but worm _heads_ don't. (I.e., the opposite of emmiegit's first idea.)

I think it's perfectly natural for ivies to be able to climb upward, btw. :)
